binary_port/src/state_request.rs

```diff
@@ -46,52 +50,66 @@ pub enum GlobalStateRequest {
         /// Dictionary item identifier.
         identifier: DictionaryItemIdentifier,
     },
+    /// Get balance by block and purse.
+    BalanceByBlock {
+        /// Block identifier, `None` means "latest block".
+        block_identifier: Option<BlockIdentifier>,
+        /// Purse identifier.
+        purse_identifier: PurseIdentifier,
+    },
+    /// Get balance by state root and purse.
+    BalanceByStateRoot {
+        /// Global state identifier, `None` means "latest block state".
+        state_identifier: Option<GlobalStateIdentifier>,
```

Comment:
that's not a bad idea, but I think we can't easily get rid of the duplication (without introducing unnecessary overhead), because these two are used in slightly different contexts:
- `resolve_state_root_hash` when we just need the state root hash and we don't need a header
- when we need the header and potentially things from inside the header as well

Reply:
after having a closer look, there is some duplication we could remove here (three branches in `resolve_state_root_hash`), but still, `GlobalStateIdentifier` is part of the public API that already existed in 1.5, so I think we probably shouldn't change it